dd-sdk-android/src/main/kotlin/com/datadog/android/core/internal/thread/ThreadPoolExecutorExt.kt

```diff
@@ -0,0 +1,26 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.android.core.internal.thread
+
+import java.util.concurrent.ThreadPoolExecutor
+import java.util.concurrent.TimeUnit
+
+internal const val MAX_SLEEP_DURATION_IN_MS = 25L
+
+internal fun ThreadPoolExecutor.waitToIdle(timeout: Long): Boolean {
+    val startTime = System.nanoTime()
+    val timeoutInNs = TimeUnit.MILLISECONDS.toNanos(timeout)
+    val sleepDurationInMs = MAX_SLEEP_DURATION_IN_MS.coerceAtMost(timeout.coerceAtLeast(0L))
+    do {
+        if (this.taskCount - this.completedTaskCount <= 0) {
+            return true
+        }
+        Thread.sleep(sleepDurationInMs)
+    } while ((System.nanoTime() - startTime) < timeoutInNs)
```

Comment:
Why are you using a while loop ? The `Thread.sleep()` uses the same timeout as the `while` check does, so it means that you'll never do more than two loops.

Reply:
Not quite...the `Thread.sleep` uses a timeout which is max `25` ms. If you give it a timeout of 2 seconds for example it will do more than 2 loops.

Reply:
well in most cases (except when the timeout is really high), `sleepDurationInMs` will be exactly equal to `timeoutInNs`. This means that you'll do one loop (or two if the stars align) and reach the exit condition of the while loop.

Reply:
Yes I see what you're saying...I am going to try to think of something more granular there or if not maybe I will use a `CountdownLatch` with a timeout.

Reply:
So I lowered the `MAX_SLEEP_DURATION` to `10` milliseconds. This will make it granular enough and it will do at least 2 loops for any timeout provided considering that you will not give a timeout less than 30ms. The reason behind the loop is to not have to wait for the whole duration of the `timeout` if the condition is met so we are waking up the thread from time to time and check the condition. In case the `timeout` is bigger this will provide better performances.